smart_open/asb.py

```diff
@@ -0,0 +1,400 @@
+        # logger.debug('reading %r bytes from %r byte-long buffer', size, len(self._current_part))
+        size = size if size >= 0 else len(self._current_part)
+        part = self._current_part.read(size)
+        self._current_pos += len(part)
+        # logger.debug('part: %r', part)
+        return part
+
+    def _fill_buffer(self, size=-1):
+        size = size if size >= 0 else self._current_part._chunk_size
+        while len(self._current_part) < size and not self._eof:
+            bytes_read = self._current_part.fill(self._raw_reader)
+            if bytes_read == 0:
+                logger.debug('reached EOF while filling buffer')
+                self._eof = True
+
+    def __str__(self):
+        return "(%s, %r, %r)" % (self.__class__.__name__, self._container.container_name, self._blob.blob_name)
+
+    def __repr__(self):
+        return "%s(container=%r, blob=%r, buffer_size=%r)" % (
+            self.__class__.__name__, self._container.container_name, self._blob.blob_name, self._current_part_size,
+        )
+
+
+class BufferedOutputBase(io.BufferedIOBase):
+    """Writes bytes to Azure Storage Blob.
+
+    Implements the io.BufferedIOBase interface of the standard library."""
+
+    def __init__(
+            self,
+            container,
+            blob,
+            client=None,  # type: azure.storage.blob.BlobServiceClient
+    ):
+        if client is None:
+            client = BlobServiceClient()
+        self._client = client
+        self._container_client = self._client.get_container_client(container)  # type: azure.storage.blob.ContainerClient
+        self._blob = self._container_client.get_blob_client(blob)  # type: azure.storage.blob.BlobClient
+        self._total_size = 0
+
+        #
+        # This member is part of the io.BufferedIOBase interface.
+        #
+        self.raw = None
+
+    def flush(self):
+        pass
+
+    #
+    # Override some methods from io.IOBase.
+    #
+    def close(self):
+        logger.debug("closing")
+        if not self.closed:
+            self._client = None
+        logger.debug("successfully closed")
+
+    @property
+    def closed(self):
+        return self._client is None
+
+    def writable(self):
+        """Return True if the stream supports writing."""
+        return True
+
+    def tell(self):
+        """Return the current stream position."""
+        return self._total_size
+
+    #
+    # io.BufferedIOBase methods.
+    #
+    def detach(self):
+        raise io.UnsupportedOperation("detach() not supported")
+
+    def write(self, b):
+        """Write the given bytes (binary string) to the Azure Storage Blob file.
+
+        There's buffering happening under the covers, so this may not actually
+        do any HTTP transfer right away."""
+
+        if not isinstance(b, _BINARY_TYPES):
+            raise TypeError("input must be one of %r, got: %r" % (_BINARY_TYPES, type(b)))
+
+        self._total_size += len(b)
+        self._blob.upload_blob(b)
```

Comment:
I'm not very sure, is the `upload_blob` from azure API buffered ready ? aka we can can `upload_blob` many times before the upload is really triggered ?

Reply:
Ok got it, it's uploaded in one time bu it handles chunk on Azure side: [doc](https://docs.microsoft.com/en-us/python/api/azure-storage-blob/azure.storage.blob.blobclient?view=azure-python#upload-blob-data--blob-type--blobtype-blockblob---blockblob----length-none--metadata-none----kwargs-)